Feat/frontend palette switcher

preda-bogdan opened the pull request:
References: Codeinwp/neve#2345

### Customizer:


### Notes:

When Automatically adjust is enabled it will use the user system preferences.
When the user toggles the switch it overwrites those with the setting that is being saved to the browser local storage.

selul commented on inc/core/dynamic_css.php:
```diff
@@ -99,7 +99,10 @@ public static function minify_css( $css ) {
-		wp_add_inline_style( 'nv-css-vars', self::minify_css(':root{' . $this->get_css_vars() . '}' ) );
+
+		$css = ':root{' . $this->get_css_vars() . '}';
+		$css .= apply_filters( 'neve_after_css_root', $css );
```
Why do you apply the same filter twice? here 👆 and here https://github.com/Codeinwp/neve/pull/2783/files#diff-1f14dcbb2fabbdeedd5ee5337cec3c42701a242214f8ee67fb7d9a6c14f8b46bR121

preda-bogdan replied:
Because `$this->get_css_vars()` is called twice once for customizer and once for frontend. Without this it does not reflect the changes inside customizer.

selul commented on header-footer-grid/templates/components/component-palette-switch.php:
```diff
@@ -0,0 +1,25 @@
+$label     = component_setting( PaletteSwitch::PLACEHOLDER_ID );
+?>
+<div class="toggle-palette">
+	<a href="<?php echo '#'; ?>" class="toggle palette-icon-wrapper">
```
the HTML needs to be Accessible ready. please validate it against some accessibility testers

preda-bogdan replied:
I've ran this through `Lighthouse Accessibility` and it does not report any issues. Is there another tool that I should be using?

selul commented on header-footer-grid/assets/scss/frontend/components/_palette-switch.scss:
```diff
@@ -0,0 +1,30 @@
+.toggle-palette {
```
this needs to be added as inline style, minified. font size, line height, letter spacing should be based on the main menu used.

selul replied:
@JohnPixle how about here, should we use the font settings of the label as the font settings of the menu?

preda-bogdan replied:
I've added the class that is being used for the primary navigation for the label so that it will use the same settings and removed the css font styles. Also added the whole style as inline minified, instead of having it compile like some of the components.

selul commented on header-footer-grid/assets/scss/frontend/components/_palette-switch.scss:
```diff
@@ -0,0 +1,30 @@
+		display: inline-flex;
+		font-size: small;
+		vertical-align: middle;
+		color: var(--nv-primary-accent);
```
please validate with @JohnPixle the color of the icon/label

JohnPixle replied:
@preda-bogdan @selul the color is correct as primary accent, thanks!

selul commented on header-footer-grid/Core/Components/PaletteSwitch.php:
```diff
@@ -0,0 +1,395 @@
+		global $wp_filesystem;
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+		WP_Filesystem();
+		$file            = NEVE_MAIN_DIR . 'header-footer-grid/assets/js/components/palette_switch.js';
```
if the file < 1KB minified, we can drop it minified as an inline script, otherwise, we can enqueue it in the footer with wp_script_add_data as async.

preda-bogdan replied:
I've included the minified `js` directly here: https://github.com/Codeinwp/neve/pull/2783/commits/0254586b62881803b499373dea5411a6fbbeb40d#diff-6f8db8b20976151ccb70279d3bf5e266944be7f0f0eb6375d020878468b7c53dR89

selul commented on header-footer-grid/Core/Customizer/Description_Control.php:
```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Name:    Header Footer Grid - Description Control
```
we already have this type of control somewhere, @abaicus used for blog typography shortcut

preda-bogdan replied:
I've removed this control, and modified `Responsive_Toggle` control from `React` to allow for description text not just links. Previously it was only allowing links to be passed.

 @abaicus If you have any input on this change let me know:  https://github.com/Codeinwp/neve/pull/2783/commits/40422d0215cbbd06f922eb2a1a17e19b5f0c7582

selul replied:
@preda-bogdan we have already a non-responsive toggle used widely in theme

selul reviewed (changes requested):
@preda-bogdan added some comments.
- please update the branch with the development
- please review the AMP compatibility
- What happens when the saved value is different from the system value? Won't be a flash initially until the js is loading?

JohnPixle commented:
Really like the screenshots so far, cant wait to see it live on the starter sites.

May I ask something UX-wise:

Assuming that the user has 3 color palettes: A, B and C.

In Customizer > Color he has Palette A set as default.
In the palette switcher, the user selects B as light palette, and C as dark.

Which palette is loaded by default for the frontend user?

preda-bogdan commented:
> Really like the screenshots so far, cant wait to see it live on the starter sites.
> 
> May I ask something UX-wise:
> 
> Assuming that the user has 3 color palettes: A, B and C.
> 
> In Customizer > Color he has Palette A set as default.
> In the palette switcher, the user selects B as light palette, and C as dark.
> 
> Which palette is loaded by default for the frontend user?

If the pallet switcher is active then for the frontend user it will display the following:

1. If the auto-adjust is enabled it will load B or C depending on the user system preference.
2. If the auto-adjust is disabled it will load B as a default.
3. All options are override by the toggle that saves the user preference to local storage.

This is the way it should work now. Do you have any feedback on this?

abaicus commented on header-footer-grid/Core/Components/PaletteSwitch.php:
```diff
@@ -0,0 +1,378 @@
+
+		$auto_adjust = get_theme_mod( $this->get_id() . '_' . self::AUTO_ADJUST, 1 );
+
+		$customizer       = get_theme_mod( 'neve_global_colors', neve_get_global_colors_default( true ) );
```
@preda-bogdan I think we should check if the palette still exists. A user might delete the palette after setting up the toggle with it in use. Maybe we should treat that case by falling back on the `darkMode` if it's not there.


@JohnPixle @selul Do we need two selectors for the palettes? Doesn't this kind of override the Global colors selected palette? I think we could do something like use the currently selected one for the light color, and only allow the user to pick the second one. Does it make sense?

preda-bogdan replied:
@abaicus thank you for pointing out this, I added some defaults for now in case the selected  color is not available anymore. I also think we can use only one palette selector to define the contrast palette and use the global defined as a base palette.

JohnPixle replied:
> @JohnPixle @selul Do we need two selectors for the palettes? Doesn't this kind of override the Global colors selected palette? I think we could do something like use the currently selected one for the light color, and only allow the user to pick the second one. Does it make sense?

Yes, it does, this is what I pointed out in the [last comment](https://github.com/Codeinwp/neve/pull/2783#issuecomment-825453098)

I believe that the currently selected palette in the customiser should always be used in the switcher. And users simply choose one of the other palettes to be used as their dark variant. I think this is what you suggesting too @abaicus  ?

abaicus replied:
@JohnPixle Yes, that is what I meant 👍🏻 I think it would make more sense 👯

selul replied:
Let's do this too, @preda-bogdan

preda-bogdan replied:
cc: @selul, @abaicus @JohnPixle 
I've removed the light selector and I use the active palette as the light one.

selul replied:
@JohnPixle can we update the Figma design regarding this change?

JohnPixle replied:
@preda-bogdan Did a quick re-work on this, please see the latest change in [figma](https://www.figma.com/file/LBW82Rm2SA5rvvLdKrRnBl/Neve-Core-UI?node-id=792%3A20988). 

We can make use of help text only in this case, and link to the color palettes.

![Screenshot 2021-04-27 at 12 27 38](https://user-images.githubusercontent.com/4193672/116219611-7790c580-a754-11eb-8e9d-d133c4a2b5bc.png)

preda-bogdan replied:
@JohnPixle I've updated the control.
![image](https://user-images.githubusercontent.com/23024731/116379810-0ae0fe00-a81c-11eb-8010-add874332ee9.png)

cc: @selul

selul commented on header-footer-grid/templates/components/component-palette-switch.php:
```diff
@@ -0,0 +1,25 @@
+$label     = component_setting( PaletteSwitch::PLACEHOLDER_ID );
+?>
+<div class="toggle-palette">
+	<a href="<?php echo '#'; ?>" class="toggle palette-icon-wrapper">
```
What's is this for? `<?php echo '#'; ?>` why not use simple href="#"

selul commented on header-footer-grid/Core/Components/PaletteSwitch.php:
```diff
@@ -0,0 +1,409 @@
+	 * @return string
+	 */
+	public function toggle_script() {
+		return '"use strict";const attributeDataName="data-neve-theme",localStorageName="neve_user_theme";function detectColorScheme(){let e="light";if(localStorage.getItem(localStorageName))"dark"===localStorage.getItem(localStorageName)&&(e="dark");else{if(!window.matchMedia)return!1;window.matchMedia("(prefers-color-scheme: dark)").matches&&(e="dark")}"dark"===e&&document.documentElement.setAttribut ...
```
this is not minified, we need to compress(uglify) the string too

selul commented on header-footer-grid/Core/Components/PaletteSwitch.php:
```diff
@@ -0,0 +1,409 @@
+
+		$customizer       = get_theme_mod( 'neve_global_colors', neve_get_global_colors_default( true ) );
+		$defined_palettes = $customizer['palettes'];
+		$active_light     = get_theme_mod( $this->get_id() . '_' . self::LIGHT_PALETTE_ID, $default_light );
```
please use the Mod class for theme mods access.

selul reviewed (changes requested):
The CI checks needs to pass too.

irinelenache commented:
@preda-bogdan 

- [ ] after i click the palette switcher, it becomes underscored until i click somewhere else, and the line looks a bit odd, i think because i changed the icon size https://prnt.sc/124s6n3
- [ ] the switcher doesn't seem to work on mobile (on my phone, the default was dark, as expected, but the switch still didn't work) https://www.loom.com/share/e850b88dcc744101afce267207d24211

selul commented:
> please review the AMP compatibility

Let's not forget about this too ☝️

preda-bogdan commented:
> > please review the AMP compatibility
> 
> Let's not forget about this too ☝️

I've checked the AMP support and made it work, one thing will miss from AMP is `localStorage` preference save since AMP does not allow localStorage access.

irinelenache commented:
@preda-bogdan for me, the palette switch still doesn't work on mobile, you can check here: 
 URL: https://buzzposition.tastewp.com/wp-admin 
 Username: admin 
 Password: WxIYJHgafU0 

Let me know if i missed something 🙏

preda-bogdan commented:
> @preda-bogdan for me, the palette switch still doesn't work on mobile, you can check here:
> URL: https://buzzposition.tastewp.com/wp-admin
> Username: admin
> Password: WxIYJHgafU0
> 
> Let me know if i missed something 🙏

It looks like you were not using the latest build. I've updated the build on your testing environment that you have provided.
It looks like it is working on mobile, can you please confirm?

Thank you!

selul commented:
@irinelenache can you confirm if the zip that you had installed is the zip attached to this PR? 

@preda-bogdan if those are different we need to evaluate why those are different and look into fixing it.

irinelenache commented:
@preda-bogdan @selul i had the latest zip that was linked to this PR, and the first issue that i mentioned was fixed, now the switcher is working on mobile too 👍

selul commented:
@irinelenache to confirm, the mobile is working fine with the zip from the PR too?

irinelenache commented:
@selul i tried again with the zip from the PR, downloaded it from here https://prnt.sc/12h5rbo and it still isn't working for me

preda-bogdan commented:
> @selul i tried again with the zip from the PR, downloaded it from here https://prnt.sc/12h5rbo and it still isn't working for me

@irinelenache I've just done the same thing using this: 
Plugin build for 9b36328089988c219855bb3220f6fe2606c22a39 is ready :bellhop_bell:!
- Download [build](https://verti-artifacts.s3.amazonaws.com/neve-feat-frontend_palette_switcher-0b27c0f3/neve.zip)

_Originally posted by @github-actions[bot] in https://github.com/Codeinwp/neve/issues/2783#issuecomment-824639591_

And it is working as expected, you can see it here: https://noiselessmint.tastewp.com/
 Admin area URL: https://noiselessmint.tastewp.com/wp-admin 
 Username: bogdanpreda 
 Password: 6TRU2uTui1E 

It looks like the build is updated, maybe it is something related to the commit wording.
![image](https://user-images.githubusercontent.com/23024731/117133997-ad195c80-adad-11eb-87ff-6a44a6faef7d.png)


Can you please check again and confirm that everything is working correctly now?
Thank you!

irinelenache commented:
@preda-bogdan now everything works fine with the last plugin build 🚀

selul commented:
@irinelenache @preda-bogdan seems like is not working as it should.

On a fresh site, after I go to customizer I get by default the black theme, even though I have the light one as base and I don't have the dark palette added to the header.

irinelenache commented:
@selul i could replicate it only on my phone, where i have the Dark mode by default, and i saw that if i add the switcher to the header, disable this option https://prnt.sc/12kq8l9 and remove the switcher from the header, the color palette will be the base one on every device

preda-bogdan commented:
@selul I've now corrected this, let me know if this approach is ok. The issue was that the filter was called even when the component is not added. 

@irinelenache You can test with default theme preferences from chrome: https://stackoverflow.com/questions/57606960/how-can-i-emulate-prefers-color-scheme-media-query-in-chrome

irinelenache commented:
@preda-bogdan tested it again and it seems to work as expected 👍 if @selul thinks the approach is ok, this is ready to merge

selul commented:
Still not working fine for me, the dark theme is by default used. Also I see the palette CSS being generated even if the component is not used https://d.pr/draw/VEQZfc/

selul commented:
@preda-bogdan the switcher is no longer working https://burytank.tastewp.com/, clicking the component does nothing

selul commented:
@preda-bogdan works good now, @irinelenache once is ready from your end we can move to ready.

selul commented:
@preda-bogdan seems like on mobile is not acting properly https://vertis.d.pr/Ms7IV5

preda-bogdan commented:
@selul I've changed the hooks so they don't overlap on mobile. Previously I added the `touch` hook so that it solved Irinel's issue on mobile. I've tested with just the `click` event and it works fine on my mobile device and on an iphone. The `touch` caused the mentioned issue.

You can quickly see the changes here: https://noiselessdoll.tastewp.com/
Let me know if this is now resolved.

Thank you!

gutoslv commented:
It's working fine on mobile now. If @selul approves it we can move to ready.
